File: app/app/src/main/java/dev/steamvault/app/ui/library/logic/Format.kt
package dev.steamvault.app.ui.library.logic

import java.util.Locale

/**
 * Bytes -> a short "N GB" string (WP 4b.4) -- Kotlin port of
 * `web/js/lib/format.js`'s `formatBytesGB` (mockup's `gb()` helper): >=100
 * GB rounds to a whole number, otherwise one decimal place. `null`/
 * non-positive bytes are never faked into a number -- callers decide what
 * to show instead (mockup rule: "a never-downloaded game shows the icon
 * alone rather than a fake dash").
 * @return `null` when there is nothing honest to print.
 */
fun formatBytesGB(bytes: Long?): String? {
    if (bytes == null || bytes <= 0L) return null
    val gb = bytes / 1_073_741_824.0
    // Locale.ROOT, not the default locale: a German (or any comma-decimal)
    // device would otherwise render "1,5 GB" -- same class of bug
    // docs/LEARNINGS.md's PowerShell section documents for locale-sensitive
    // number formatting ("force InvariantCulture"), here on the JVM side.
    val formatted = if (gb >= 100) {
        String.format(Locale.ROOT, "%.0f", gb)
    } else {
        String.format(Locale.ROOT, "%.1f", gb)
    }
    return "$formatted GB"
}

File: web/js/lib/format.js
/**
 * Small formatting helpers shared by the library and (later) downloads
 * views (WP 4a.3). Pure — unit-tested in web/tests/format.test.js.
 */

/**
 * Bytes -> a short "N GB" string, mirroring the mockup's `gb()` helper
 * (>=100 GB rounds to a whole number, otherwise one decimal place).
 * `null`/`undefined`/non-positive bytes are never faked into a number —
 * callers decide what to show instead (mockup rule: "a never-downloaded
 * game shows the icon alone rather than a fake dash").
 * @param {number | null | undefined} bytes
 * @returns {string | null} `null` when there is nothing honest to print.
 */
export function formatBytesGB(bytes) {
  if (typeof bytes !== "number" || !Number.isFinite(bytes) || bytes <= 0) return null;
  const gb = bytes / 1_073_741_824;
  return (gb >= 100 ? gb.toFixed(0) : gb.toFixed(1)) + " GB";
}

/**
 * Bytes -> a short "N GB" string, WITHOUT collapsing a genuine zero into
 * "nothing to print" the way {@link formatBytesGB} deliberately does
 * (that helper's own test file: "null/undefined/zero/negative never
 * fabricate a number" — a design choice for the library tile badge, where
 * 0 means "never downloaded" and the mockup rule is to show the icon alone).
 * The rail foot (WP 4e.6) needs the OPPOSITE rule for the same input shape:
 * "0 bytes free" is a real, alarming fact (the disk is full) that must
 * render, not a `null`-shaped "we don't know" — collapsing the two would
 * hide a full disk exactly as silently as it hides "no poll has landed yet",
 * which is the ambiguity this whole work package exists to remove. Only
 * `null`/`undefined`/negative/non-finite input (never a valid byte count)
 * still maps to `null` — zero is a legitimate, printable value here.
 * @param {number | null | undefined} bytes
 * @returns {string | null} `null` only for input that could never be a real
 *   byte count; a real zero prints as `"0.0 GB"`.
 */
export function formatBytesGBOrZero(bytes) {
  if (typeof bytes !== "number" || !Number.isFinite(bytes) || bytes < 0) return null;
  const gb = bytes / 1_073_741_824;
  return (gb >= 100 ? gb.toFixed(0) : gb.toFixed(1)) + " GB";
}

/**
 * An ISO-8601 timestamp (`jobs.created_at`/`started_at`/`finished_at`, ...)
 * -> a locale-formatted string for display (WP 4a.5, Downloads history
 * rows). `null`/`undefined`/unparseable input never fabricates a time —
 * same "nothing honest to print" posture as {@link formatBytesGB}.
 * @param {string | null | undefined} iso
 * @returns {string} a display string, or "—" when there is nothing to show.
 */
export function formatTimestamp(iso) {
  if (!iso) return "—";
  const d = new Date(iso);
  if (Number.isNaN(d.getTime())) return "—";
  return d.toLocaleString();
}
